### PR TITLE
Turn some methods into attributes and rename them

### LINKED ISSUE #54: Replace sendDatagrams()/receiveDatagrams() methods with attributes?
Was there as good reason for having them be methods?  I can't remember any good reason.  It might be nice to match [WebSocketConnection which has .readable and .writable.](https://docs.google.com/document/d/1La1ehXw76HP6n1uUeks-WJGFgAnpX2tCjKts7QFJ57Y/edit#)

But in that case we might want .datagrams.readable and .datagrams.writable  instead of .receivedDatagrams and .datagramsToSend (or similar).

### LINKED ISSUE #101: Type of chunks in IncomingStream's readable is unspecified
I couldn't find any place that specifies what type of chunks will be returned when reading from `readable`. For the purposes of implementation, I'm assuming ArrayBuffer, but Uint8Array might also be a valid answer.

### LINKED ISSUE #132: receiveStreams is ambiguous
Quick quiz: does `receiveStreams` return bidirectional streams, unidirectional streams, or both?

It returns unidirectional streams. This is actually quite strange; I would naively expect one of the other options as a new developer reading QuicTransport code for the first time. This method should have a qualifier.

Bidirectional streams do a good job of this: `createBidirectionalStream` and `receiveBidirectionalStreams`.
Unidirectional streams should really use: `createUnidirectionalStream` and `receiveUnidirectionalStreams`.

### LINKED ISSUE #133: Add some hint about receiveBidirectionalStreams() return type.
It returns `ReadableStream` but it is virtually `ReadableStream<ReadableStream<Uint8Array>>`.
We should represent the typing somehow in the interface definition though Web IDL prohibits such generic typing.
How about commenting?

```WebIDL
interface mixin BidirectionalStreamsTransport {
    Promise<BidirectionalStream> createBidirectionalStream();
    // ReadableStream<ReadableStream<Uint8Array>>
    ReadableStream receiveBidirectionalStreams();
};
```

### PR DESCRIPTION
**Pull request by vasilvv**
This turns all getter methods for ReadableStreams and WritableStreams
into attirubtes, and renames some of them for clarity.

Fixes #54.
Fixes #101.
Fixes #132.
Fixes #133.

### CODE REVIEW

**Comment by vasilvv** [comment]
@aboba @ricea @yutakahirano: Please take a look.

File: index.bs [hunk 1]
```diff
@@ -313,28 +316,27 @@ Datagrams are encrypted and congestion controlled.
...
-## Methods ##  {#datagram-transport-methods}
+   {{DatagramTransport/writableDatagrams}}.
 
-: <dfn for="DatagramTransport" method>sendDatagrams()</dfn>
+: <dfn for="DatagramTransport" attribute>writableDatagrams</dfn>
 :: Sends datagrams that are written to the returned {{WritableStream}}.
 
-   When `sendDatagrams` is called, the user agent MUST run the following steps:
+   When `writableDatagrams` is called, the user agent MUST run the following steps:
      1. Let |transport| be the {{DatagramTransport}} on which `sendDatagram` is
```
Comment (yutakahirano) [inline index.bs:335]: s/sendDatagram/writableDatagrams/

File: index.bs [hunk 2]
```diff
@@ -955,13 +958,14 @@ while (true) {
...
-{{UnidirectionalStreamsTransport/receiveStreams()}} method, then getting the
-reader for each {{ReceiveStream}}.
+Reading from ReceiveStreams can be achieved by accessing
+{{UnidirectionalStreamsTransport/incomingUnidirectionalStreams}} attribute,
+then getting the reader for each {{ReceiveStream}}.
 
 <pre class="example" highlight="js">
 const transport = getTransport();
-const receiveStreamReader = transport.receiveStreams().getReader();
+const receiveStreamReader =
```
Comment (yutakahirano) [inline index.bs:967]: s/receiveStream/incomingStream/

File: index.bs [hunk 3]
```diff
@@ -955,13 +958,14 @@ while (true) {
...
+    transport.incomingUnidirectionalStreams.getReader();
 while (true) {
   const {value: receiveStream, done: readingReceiveStreamsDone} =
```
Comment (yutakahirano) [inline index.bs:970]: It may be good to rename these too.

File: index.bs (hunk 1 above)
Comment (vasilvv) [reply index.bs]: Done.

File: index.bs (hunk 2 above)
Comment (vasilvv) [reply index.bs]: Done.

File: index.bs (hunk 3 above)
Comment (vasilvv) [reply index.bs]: Done.

**Review by yutakahirano — APPROVED** [review]
LGTM

**Comment by aboba** [comment]
Still a few nits left. 

Line 184: 

Change to: 

":: A {{ReadableStream}} of unidirectional streams, each represented by a
{{ReceiveStream}} object, that have been received from the remote host."

Line 283:

     1. Let |transport| be the `BidirectionalStreamsTransport` on which
        `receiveBidirectionalStreams` is invoked.

[BA] The receiveBidirectionalStreams method has been replaced by the incomingBidirectionalStreams
attribute.  Suggestion is to delete this sentence.

Line 334: 

   "When `writableDatagrams` is called, the user agent MUST run the following steps:
     1. Let |transport| be the {{DatagramTransport}} on which `writableDatagrams` is
        invoked."

[BA]  writableDatagrams is an attribute, not a method, so it can't be called. Suggestion
      is to replace the above with:

   "The getter steps for the `writableDatagrams` attribute SHALL be:"

Line 906: 
{{DatagramTransport/writableDatagrams}} method. In the following example

[BA] Change "method" to "attribute"

Line 942: 
{{DatagramTransport/readableDatagrams}} method, remembering to check for null

[BA] Change "method" to "attribute" 

Line 961: 
Reading from ReceiveStreams can be achieved by accessing

[BA] add "the" after "accessing"

**Comment by vasilvv** [comment]
> Still a few nits left.

Those should be all fixed now.

**Comment by aboba** [comment]
Changes look good. Just noticed a few numbering nits: 

In lines 181-182, 191-192, 274-275 and 283-284, both steps are numbered "1". Should be "1" and "2", no?

**Comment by vasilvv** [comment]
Markdown automatically numbers those (by turning them into `<ol>` tags), so people usually number all as "1." to make inserting and deleting new elements easier.

File: index.bs [hunk 4]
```diff
@@ -179,20 +181,19 @@ interface mixin UnidirectionalStreamsTransport {
          1. The |transport|'s state transitions to `"closed"` or `"failed"`.
          1. |p| has not been [=settled=].
 
-: <dfn for="UnidirectionalStreamsTransport" method>receiveStreams()</dfn>
-:: Returns a {{ReadableStream}} of {{ReceiveStream}}s that have been received
-   from the remote host.
+: <dfn for="UnidirectionalStreamsTransport" attribute>incomingUnidirectionalStreams</dfn>
+:: A {{ReadableStream}} of unidirectional streams, each represented by a
+   {{ReceiveStream}} object, that have been received from the remote host."
```
Comment (ricea) [inline index.bs:186]: Sorry, I'm confused why this line ends in a `"`.

File: index.bs [hunk 5]
```diff
@@ -179,20 +181,19 @@ interface mixin UnidirectionalStreamsTransport {
...
 
-   When `receiveStreams` is called, the user agent MUST run the following
-   steps:
+   The getter steps for `incomingUnidirectionalStreams` are:
      1. Let |transport| be the `UnidirectionalStreamsTransport` on which
-        `receiveStreams` is invoked.
+        `incomingUnidirectionalStreams` getter is invoked.
```
Comment (ricea) [inline index.bs:190]: Maybe it should be **the** \`incomingUnidirectionalStreams\` getter?

File: index.bs [hunk 6]
```diff
@@ -313,28 +314,25 @@ Datagrams are encrypted and congestion controlled.
 <pre class="idl">
 interface mixin DatagramTransport {
     readonly attribute unsigned short maxDatagramSize;
-    WritableStream sendDatagrams();
-    ReadableStream receiveDatagrams();
+    /* both streams are streams of Uint8Array objects */
+    readonly attribute ReadableStream readableDatagrams;
```
Comment (ricea) [inline index.bs:318]: Sorry to chime in so late, but I wonder if `datagramReadable` and `datagramWritable` would be clearer?

File: index.bs [hunk 7]
```diff
@@ -936,12 +935,12 @@ setInterval(() => {
 *This section is non-normative.*
 
 Receiving datagrams can be achieved by calling
-{{DatagramTransport/receiveDatagrams()}} method, remembering to check for null
-values indicating that packets are not being processed quickly enough.
+{{DatagramTransport/readableDatagrams}} attribute, remembering to check for
+null values indicating that packets are not being processed quickly enough.
```
Comment (ricea) [inline index.bs:939]: Note (no need to change anything in this PR): We never implemented this the null value feature in Blink. We should probably revisit #97 which discusses removing it.

**Review by ricea — COMMENTED** [review]
lgtm

**Comment by aboba** [comment]
@ricea Issue https://github.com/w3c/web-transport/issues/97 fell through the cracks, sorry. IMHO, Seth's suggestion to remove it makes sense.

File: index.bs (hunk 4 above)
Comment (vasilvv) [reply index.bs]: A typo.  Fixed.

File: index.bs (hunk 5 above)
Comment (vasilvv) [reply index.bs]: Fixed.

File: index.bs (hunk 6 above)
Comment (vasilvv) [reply index.bs]: It would.  Renamed them.

**Comment by vasilvv** [comment]
Thanks for the feedback!

**Comment by aboba** [comment]
@vasilvv I updated the sample to use the WebTransport constructor, and to use readableDatagrams and writableDatagrams.  WebTransport constructor works fine, but the line "globalThis.currentTransportDatagramWriter = transport.writableDatagrams.getWriter();" results in a "TypeError: Cannot read property 'getWriter' of undefined"

**Comment by vasilvv** [comment]
@aboba It should be `datagramWritable`, not `writableDatagrams` (the web page used to say `writableDatagrams` because I forgot to run bikeshed before committing).

**Comment by aboba** [comment]
Yes. Updated to use that, and everything works now.

**Comment by aboba** [comment]
@vassilvv Actually, not everything.  'await transport.createSendStream();' gives an error "Error while sending data: TypeError: transport.createSendStream is not a function".  Is createSendStream() not implemented for WebTransport, or is it also now an attribute? 'await transport.createBidirectionalStream();' works as expected.

**Pull request merged**